src/core/service.c

```diff
@@ -5229,38 +5229,21 @@ static int service_live_mount(Unit *u,
 
         r = service_arm_timer(s, /* relative= */ true, s->timeout_start_usec);
         if (r < 0) {
-                log_unit_warning_errno(u, r, "Failed to install timer: %m");
-                sd_bus_error_set_errnof(
```

Comment:
Please leave all of these in place - we need good and intelligible error strings to be returned to the callers, with as much detail as possible, as the usual error numbers are unintelligible

Reply:
As mentioned, they are useless. First of all, this really shouldn't fail in the first place. And if pid1 is borked, it's our internal state that the caller cannot interfere in any means.

See `bus_unit_method_freezer_generic` for example. We selectively report known errors to caller so that it could be handled. Otherwise we just return errno and log.

Reply:
All other code follows this code, including `dbus-service` transient property setters.

Reply:
Yes in general we are really really really bad at returning good errors, it's one of the major issues we have, it's very often impossible to figure out what an issue is without someone who can run things through a debugger. This is a problem to solve though, and we need to do better

Reply:
Quite frankly, let's not conflate two types of errors - the workable ones and theoretical ones. We should always generate well-formed errors for the errors that the caller should handle, but the latter is simply not actionable. You'd definitely need a debugger to track down "theoretical errors" when they pop up in real life, because they reflect extremely low-level stuff.

Reply:
If it cannot happen then it should be an assert, not a return with error. Anytime an IPC call returns with a random -EWHATEVER and nothing else, everyone is left scratching their heads because they are just incomprehensible, and instead support tickets get created, which are even more incomprehensible, because -EWHATEVER can happen in thousands of different places and nobody has any clue and it costs an inordinate amount of time and money just to track it down. We can do better than that, and I know this is not the case in most places, but again that's an issue to solve, not a reason to do worse

Reply:
I agree with @bluca's claim.
The error is actually propagated to the caller. So, we should provide details about error, even if users cannot take any action on the failure. At least, users can open an issue with the error message.

Please keep them.

Reply:
I'd side with @YHNdnzj on this: we should be careful with which errors we propagate to clients and which ones we don't. Errors in the user's provided data, definitely need to be returned to the user. Errors during execution of the operation that are "expected" similar. But "unexpected" errors, that indicate a logic error on our side, and just reveal internal details and bugs are better placed in PID1's logs I am sure, as they are not something a client reasonably could do anything about it.

I think key really is that errors that are actionable to users should be returned in a way actionable to them. But internal errors shouldn't, and system logs are the better place for them.

Of course the line between these two areas is blurry, but inability to install a timer in my PoV is clearly on the side of things clients couldn't reasonably do anything about.

Reply:
If it's a logic error that cannot happen then make it an assert, not a return. Returning intelligible errors to API calls is not just about what a client can do, it's also about the client owners being able to report sensible and intelligible stuff to us, instead of "thing failed with -EWTF please fix it" which is what happens now, and it sucks massively

Reply:
I'd preserve the error for forking off process. The timer and watch pidref ones I sincerely don't see a point, as @poettering also agrees.

Reply:
If an error is returned to an IPC call, then there is no drawback in also returning a readable string together with -EWTF, it's always strictly better to do so, as it's additive. If an error cannot happen by construction, then it should be an assert.